### PR TITLE
initial rabbitmq

### PR DESCRIPTION
**Pull request by basz**
i hope this is useful and make some sense.

the idea is to integrate it (someday) into the professor-do applications

my hope is this could be extended such that one could use their own certificates and provided a default configuration script (via `rabbitmq add_user blah blah` or a definitions file)

### CODE REVIEW

**Review by prolic — APPROVED** [review]
LGTM

File: rabbitmq/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ENV RABBITMQ_SSL_CERTFILE ${RABBITMQ_SSL_CERTFILE:-"/etc/rabbitmq/ssl/localhost.crt"}
+ENV RABBITMQ_SSL_KEYFILE ${RABBITMQ_SSL_KEYFILE:-"/etc/rabbitmq/ssl/localhost.key"}
+ENV RABBITMQ_SSL_VERIFY ${RABBITMQ_SSL_VERIFY:-"verify_none"}
+ENV RABBITMQ_SSL_FAIL_IF_NO_PEER_CERT ${RABBITMQ_SSL_FAIL_IF_NO_PEER_CERT:-"false"}
+
+# Without this erlang argument you will not be able to connect over TLS using the rabbitctl cli tool
+ENV RABBITMQ_CTL_ERL_ARGS ${RABBITMQ_CTL_ERL_ARGS:-"-proto_dist inet_tls"}
+
+# Set up the application directory
+VOLUME ["/app"]
```
Comment (sandrokeil) [inline rabbitmq/Dockerfile]: @basz Why do we need to mount the source folder to the RabbitMQ Docker container?

**Comment by basz** [comment]
At this time we don't...

Might it be useful if want want to customize the configuration? Like provide certificates or user setup?

Answering myself; I believe now that should be done in an extending image. I'll remove it

File: rabbitmq/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# RabbitMQ Docker files with Alpine Linux
+
+> These images are only for development environments
+
+## Build upon
```
Comment (bweston92) [inline rabbitmq/README.md]: s/Build/Built/

File: rabbitmq/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+The [official](https://hub.docker.com/_/rabbitmq/) rabbitmq docker images, specifically alpine with the management plugin enabled.
+
+## What does this image do?
+
+- Configures rabbitmq to use TLS with a self-signed localhost certificate. 
+- Enables the management plugin.
+- Sets a default `prooph` user with password `prooph`.
```
Comment (bweston92) [inline rabbitmq/README.md]: It does? I'm seeing it create the user `admin`.

https://github.com/basz/docker-files/blob/9707cb2bfee3210669a1832d7a85cd492244a952/rabbitmq/Dockerfile#L7

File: rabbitmq/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+- Mounts /app as application folder.
```
Comment (bweston92) [inline rabbitmq/README.md]: You've removed this I believe.

**Comment by codeliner** [comment]
I'm wondering if we should enable [websocket plugin](https://www.rabbitmq.com/web-stomp.html) by default? Or another image extending this one?

Update: I've voted for this image because the feature is really nice for ES backends and I'd like to use it in prooph|micro-do. Furthermore, our docker images are for dev environments so "batteries included" is my favorite to keep it simple ;)

**Comment by basz** [comment]
No problem as I have figured it out already.

I would say another image might be appropriate.

Let's get some votes

**Comment by basz** [comment]
We should add it on this image

**Comment by basz** [comment]
We should add it in an other image

**Comment by sandrokeil** [comment]
I would prefer to create the prooph Docker images for `development` and keep it simple. It makes no sense to make anything customizable for others IMHO, because if you go to production, you should always build your own Docker images from official builds or from scratch.

The prooph Dockerfiles can help to figure out what lib etc. is needed for an extension. That's enough. ;-)

**Comment by basz** [comment]
I agree with sandrokeil in general - but the web stomp could be useful for the prophessor-do, prophessor-do micro and the event store gui...

**Comment by codeliner** [comment]
looks like we have winner :tada:

**Comment by basz** [comment]
> we have winner 

yeah, except that "no problem" statement... :-)

problem is that the default image creates an /etc/rabbitmq/rabbitmq.config file that contains only the basics settings (only rabbit and management). Since rabbitmq doesn't support a /etc/rabbitmq/conf.d directory it means that when you wish to add other plugin configuration a specialised /etc/rabbitmq/rabbitmq-custom.config should be provided. In my experiments I did that by overriding an ENV var so the config points to a different file which is prepared and copied into the image. Unfortunately then the configuration becomes 'static'.

Does anyone see a better way around this?

**Comment by sandrokeil** [comment]
@basz Have you tried to use a named volume? So we could copy the files to the rabbitmq container and after a restart, the config should be loaded. It's dirty but could be a workaround.

I would put all the stuff, which is needed for proophessor-do, in this Docker image. So we can copy the config files like we do with the PHP Docker images. I guess, that's the simplest solution.

**Comment by basz** [comment]
i haven't looked into named volumes. is that the --volume-from flag?

> like we do with the PHP Docker images

you mean keep an configuration file for rabbitmq in the proophessor-do repo?

**Comment by sandrokeil** [comment]
> you mean keep an configuration file for rabbitmq in the proophessor-do repo?

No, copy the configuration file for proophessor-do directly in this image or do we need multiple different configs?

**Comment by codeliner** [comment]
What is missing for this PR? Do we have a solution for the config now? Would be great to have the image available soon.

**Comment by basz** [comment]
I think I get what I need to do. Will do soon...

**Comment by basz** [comment]
@codeliner this runs... I haven't actually tested the web stomp plugin yet... no time

**Comment by codeliner** [comment]
great work @basz 
@sandrokeil can you do a last check?
LGTM

**Review by codeliner — APPROVED** [review]

**Comment by basz** [comment]
hi @sandrokeil curious as to what your thoughts here...

File: rabbitmq/test/temp-queue.html [hunk 5]
```diff
@@ -0,0 +1,103 @@
+<!DOCTYPE html>
+<html><head>
+  <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.6.2/jquery.min.js"></script>
+  <script src="//cdn.jsdelivr.net/sockjs/1.0.3/sockjs.min.js"></script>
+  <script src="stomp.js"></script>
```
Comment (sandrokeil) [inline rabbitmq/test/temp-queue.html:5]: @basz Should we use also a CDN for `stomp.js` ?

Comment (codeliner) [reply rabbitmq/test/temp-queue.html:5]: Looks like stomp.js is not available on a CDN: https://github.com/jmesnil/stomp-websocket

**Comment by codeliner** [comment]
LGTM so I merge it. Need it for event machine and PR is already open for too long. Thx for the great work @basz We will use the rabbit container next week in a two days workshop ;)

**Pull request merged**